### DISCUSSION TITLE
About "Save CTRLR state"  (one more time...)

### DISCUSSION THREAD

**Discussion by FabQi** [RomanKubiak/ctrlr#616] [Q&A]
Hello 

How can I use "Save CTRLR state" (in file menu) from lua script ?

Regards :-)
Fabrice

**Comment by dobo365** [comment]
Hi Fabrice,
   at a first step, please look at my post in https://ctrlr.org/forums/topic/global-persistent-variables/
Basically, you can create 2 methods ReadStateData and SaveStateData that you can attach to the Panel properties.
At each panel close / open, the respective method is fired and by using that you can save your own variables and get them back at next panel opening.
Just ask if you need more info ;-)
/Dominique

**Reply by FabQi** [reply]
Thank you but...
Not the answer I need, because of my wrong question I think. Sorry for my english...

Reformulation :

I want to triger/call "Save CTRLR state" from a lua script (like if the user click on it in the menu)
How can do it ?

**Reply by dobo365** [reply]
You write the SaveStateData method and I guess you just call it from the press of a button.
You are not obliged to attach it to the panel property.
But again, each time you close a restricted standalone panel, its State data is saved. Same when the VST instance is closed.

```
--
-- Called when data needs saving
-- StateData is saved when the user press Ctrl+Alt+S (Save Ctrlr StateData in the menu)
-- or when quitting Ctrlr while leaving the panel open.
-- At that time all modulators indicated as Dynamic are saved but by using this method 
-- it is possible to also save values of variables that can then be read when loading the panel
--
SaveStateData = function(--[[ ValueTree --]]stateData)

	-- Saving the last file loaded/saved so it can be reloaded at panel load
	-- This is needed to have all uiLabels restored
	-- If a uiLabel is set to Dynamic instead of Static then it is its value - 0 - and not its uiLabelText property that is saved
	-- Therefore, obliged to keep all uiLabels as Static and to reload the last file
	stateData:setProperty ("LastFileFullPath", sLastFileFullPath, nil)
	stateData:setProperty ("LastFileName", sLastFileName, nil)
	stateData:setProperty ("RootFolder", currentRootFolder, nil)

end
```

Ha but now I understand, you also want to save the actual StateData, not only the additional variables.
Hmmm. Good question. I don't know how to get the StateData value tree as a variable...
It is maybe a panel property. Check panel:getRestoreState(). It is maybe giving the StateData value tree. I didn't try it.
@damiensellier; @dnaldoog: any idea on that one?

**Reply by FabQi** [reply]
Hello

I'm using Ctrlr 5.3.201 (32bit) on Windows 7 Fr

about : "`each time you close a restricted standalone panel, its State data is saved`"

That should be enough for me but I think it's not working here

Is there a way to check it in the Ctrlr 'hiden file' (or other tips) ?

**Reply by damiensellier** [reply]
Not sure 100% but I recall it stays in ctrlr.settings file when you quit the standalone app without closing the panel the stateData should be in there at the bottom of the xml.
It's supposed to be stored in your xml settings in the DAW session file as VST/AU.

**Reply by dobo365** [reply]
_Salut Fabrice, tu peux me contacter sur sunny.synths@gmail.com pour continuer en français si tu as plus facile. Je mettrai les commentaires ici ;-)_

As I said, the state data is saved in Ctrlr.settings when you are using the Ctrlr editor OR in the your_panel_name.settings file in the temp folder of your panel for a restricted standalone instance.
It is not clear for me where the settings are stored for a VST instance. I know it is within the DAW project but I don't know how to see them.

So, Fabrice, to see the settings for a restricted instance, you go (o Windows) to Users/your_user_name/AppData/Roaming/your_panel_name folder to find the your_panel_name.settings file that you can open in Notepad.
If you have added your own state data, it will be at the end of the file.
In that file you see the names and values of each modulator, all panel settings...

```
PROP   CTRLR_WINDOW_STATE 315 139 1403 760 CTRLR_FILTER_STATE <?xml version="1.0" encoding="UTF-8"?> 
<panelState panelVersionMajor="0" panelVersionMinor="15" time="1699133171020" ctrlrShutdownDelay="512" ctrlrMaxExportedVstParameters="64" ctrlrEditorBounds="0 0 1403 760" panelMidiPauseIn="0" panelMidiPauseOut="0" panelMidiGlobalDelay="0" panelMidiMatchCacheSize="32" panelMidiProgramCalloutOnprogramChange="0" panelMidiSendProgramChangeOnLoad="0" panelMidiBankMsb="0" panelMidiBankLsb="0" panelMidiProgram="0" panelMidiInputThreadPriority="7" panelMidiRealtimeIgnore="1" panelMidiThruD2HChannelize="0" panelMidiThruD2H="0" panelMidiThruD2DChannelize="0" panelMidiThruD2D="0" panelMidiThruH2DChannelize="0" panelMidiThruH2D="0" panelMidiThruH2HChannelize="0" panelMidiThruH2H="0" panelMidiOutputChannelHost="1" 
...
<value name="Env1Attack" value="0"/><value name="Env1Release" value="0"/><value name="Env2Attack" value="0"/><value name="Env2Decay" value="0"/><value name="Env2Sustain" value="999"/><value name="Env2Release" value="0"/><value name="FilterEnvAmount" value="500"/><value name="Env1Decay" value="0"/><value name="Env1Sustain" value="0"/><value name="PolyModOscBAmount" value="0"/><value name="GlideRate" value="0"/><value name="FilterResonance" value="344"/><value name="FilterCutoff" value="711"/><value name="UnisonDetune" value="0"/><value name="NoiseLevel" value="0"/><value name="Volume" value="500"/>
...
<panelCustomData LastFileFullPath="" LastFileName="Pro-800 Init Program" LastBank="" LastProgram="A35" LastProgramNumber="0" LastLoadType="init" LastProgramName="Hello!" Disk1FullPath="D:\Sound library\Ctrlr\Pro-800\Sysex\" Disk1FileName="InitBank 2.syx" Disk2FullPath="" Disk2FileName="" Synth1FullPath="D:\Sound library\Ctrlr\Pro-800\Sysex\" Synth1FileName="PRO-800 FACTORY BANK A_New.syx" Program="INIT PROG." ProgramName="Hello!" Category="[Not defined]" Description="Init program description" Author="?" Date="?"/>
</panelState> 

```

**Comment by damiensellier** [comment]
I just checked in a cubase .cpr project file and the statedata are saved as xml like this for a ctrlr instance : 

Line 20578 

`<panelState panelVersionMajor="1.0" panelVersionMinor="1.0" time="1644983504893" ctrlrShutdownDelay="512" ctrlrMaxExportedVstParameters="64" ctrlrEditorBounds="0 0 1204 724" panelMidiPauseIn="0" panelMidiPauseOut="0" panelMidiGlobalDelay="0" panelMidiMatchCacheSize="32.0" panelMidiProgramCalloutOnprogramChange="0" panelMidiSendProgramChangeOnLoad="0" panelMidiBankMsb="0" panelMidiBankLsb="0" panelMidiProgram="0" panelMidiInputThreadPriority="7.0" panelMidiRealtimeIgnore="1" panelMidiThruD2HChannelize="0" panelMidiThruD2H="0" panelMidiThruD2DChannelize="0" panelMidiThruD2D="0" panelMidiThruH2DChannelize="0" panelMidiThruH2D="0" panelMidiThruH2HChannelize="0" panelMidiThruH2H="0" panelMidiOutputChannelHost="1" panelMidiOutputToHost="0" panelMidiInputChannelHost="1" panelMidiInputFromHost="0" panelMidiOutputDevice="Yamaha SPX900" panelMidiOutputChannelDevice="1" panelMidiControllerDevice="-- None" panelMidiControllerChannelDevice="0" panelMidiInputDevice="Yamaha SPX900" panelMidiInputChannelDevice="0" panelMidiSnapshotShowProgress="0" panelMidiSnapshotDelay="10.0" panelMidiSnapshotAfterProgramChange="0" panelMidiSnapshotAfterLoad="0" uiLuaConsoleInputRemoveAfterRun="1" ctrlrUseEditorWrapper="0" ctrlrLogOptions="32" ctrlrAutoSaveInterval="300" ctrlrAutoSave="1" ctrlrOverwriteResources="1" ctrlrLuaDisabled="0" ctrlrMidiMonOutputBufferSize="8192" ctrlrMidiMonInputBufferSize="8192" ctrlrVersionCompressed="0" ctrlrVersionSeparator="_" ctrlrLuaDebug="0" ctrlrLogToFile="0"><value name="PreFxParam01" value="0"/><value name="PreFxParam02" value="10"/><value name="PreFxParam03" value="15"/><value name="PreFxParam04" value="6"/><value name="PreFxParam05" value="0"/><value name="PreFxParam06" value="21"/><value name="PreFxParam07" value="17"/><value name="PreFxParam08" value="3"/><value name="PreFxParam09" value="7"/><value name="PreFxParam10" value="0"/><value name="PreFxParam11" value="-35"/><value name="PreFxParam12" value="0"/><value name="PreFxParam20" value="6"/><value name="PreFxParam21" value="0"/><value name="PreFxParam22" value="51"/><value name="PreFxParam23" value="-35"/><value name="PreFxParam24" value="7"/><value name="PreFxParam19" value="19"/><value name="PreFxParam18" value="24"/><value name="PreFxParam17" value="1"/><value name="PreFxParam16" value="5"/><value name="PreFxParam15" value="15"/><value name="PreFxParam14" value="12"/><value name="PreFxParam13" value="0"/><value name="PostFxParam01" value="100"/><value name="PostFxParam02" value="130"/><value name="PostFxParam03" value="0"/><value name="PostFxParam04" value="0"/><value name="MainFxParam01" value="23"/><value name="MainFxParam02" value="5"/><value name="MainFxParam03" value="5"/><value name="MainFxParam04" value="300"/><value name="MainFxParam05" value="0"/><value name="MainFxParam06" value="18"/><value name="MainFxParam07" value="6"/><value name="MainFxParam08" value="5"/><value name="MainFxParam09" value="250"/><value name="MainFxParam10" value="100"/><value name="MainFxParam11" value="18"/><value name="MainFxParam12" value="25"/><value name="MainFxParam20" value="0"/><value name="MainFxParam21" value="3"/><value name="MainFxParam22" value="0"/><value name="MainFxParam23" value="0"/><value name="MainFxParam24" value="0"/><value name="MainFxParam19" value="2"/><value name="MainFxParam18" value="12"/><value name="MainFxParam17" value="10"/><value name="MainFxParam16" value="0"/><value name="MainFxParam15" value="3"/><value name="MainFxParam14" value="1"/><value name="MainFxParam13" value="50"/><value name="MainFxParam25" value="2"/><value name="MainFxParam26" value="0"/><value name="MainFxParam27" value="0"/><value name="MainFxParam28" value="0"/><value name="MainFxParam29" value="0"/><value name="MainFxParam30" value="0"/><value name="MainFxParam31" value="0"/><value name="MainFxParam32" value="0"/><value name="MainFxParam33" value="0"/><value name="MainFxParamOutLvl1" value="130"/><value name="MainFxParamMix1" value="100"/><value name="MainFxParamMix2" value="0"/><value name="PreFxParamType" value="1"/><value name="MainFxParam01s" value="1250"/><value name="MainFxParam02s" value="8"/><value name="MainFxParam03s" value="1"/><value name="MainFxParam04s" value="0"/><value name="MainFxParam05s" value="651"/><value name="MainFxParam06s" value="35"/><value name="MainFxParam07s" value="-8"/><value name="MainFxParam08s" value="200"/><value name="MainFxParam09s" value="0"/><value name="MainFxParam10s" value="0"/><value name="MainFxParam11s" value="1"/><value name="MainFxParam12s" value="0"/><value name="MainFxParam13s" value="0"/><value name="MainFxParam14s" value="0"/><value name="MainFxParam15s" value="0"/><value name="MainFxParam16s" value="0"/><value name="MainFxParam17s" value="-70"/><value name="MainFxParam18s" value="150"/><value name="MainFxParam19s" value="0"/><value name="MainFxParam20s" value="0"/><value name="MainFxParam21s" value="0"/><value name="MainFxParam22s" value="0"/><value name="MainFxParam23s" value="-70"/><value name="MainFxParam24s" value="150"/><value name="MainFxParam25s" value="0"/><value name="MainFxParam26s" value="0"/><value name="MainFxParam27s" value="0"/><value name="MainFxParam28s" value="0"/><value name="MainFxParam29s" value="0"/><value name="MainFxParam30s" value="0"/><value name="MainFxParam31s" value="0"/><value name="MainFxParam32s" value="0"/><value name="MainFxParam33s" value="0"/><value name="BtnUtility" value="0"/><value name="BtnRecall" value="0"/><value name="BtnStore" value="0"/><value name="BtnPre" value="0"/><value name="BtnMain" value="1"/><value name="BtnPost" value="0"/><value name="GrpMainMarkupLine02" value="0"/><value name="CreditPix" value="0"/><value name="RcxToSpxPix" value="0"/><value name="SyxToSpxPix" value="0"/><value name="SpxToSyxPix" value="0"/><value name="SpxMidiPix" value="0"/><value name="MainFxParamOutLvl2" value="0"/><panelCustomData UserPrgrmName="" LastUsedProgram="f0 43 00 7e 02 0a 4c 4d 20 20 38 36 33 36 4d 01 02 00 05 02 04 05 05 06 03 01 02 00 02 00 02 00 04 08 04 01 04 0c 04 0c 02 00 02 00 02 00 02 00 00 00 01 0a 00 00 00 06 00 00 00 05 00 01 02 0c 00 00 00 00 00 00 03 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 03 02 00 00 00 01 00 00 00 04 00 00 00 00 0f 0f 0b 0a 00 00 09 06 00 00 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00 01 00 00 00 0b 00 00 00 07 00 00 03 03 00 00 00 07 00 00 01 00 00 04 00 01 00 01 01 09 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 06 04 00 00 08 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 03 0e 08 00 00 03 0c 00 00 00 00 00 03 0e 08 00 00 00 01 24 f7" SaveStateDisplaySizeIndex="0.0"/></panelState>`

So it's compliant to VST standard for VST presets and the custom datas are in the **panelCustomData** key as I programmed it to store "LastUsedProgram" as a string in my saveState LUA script.

Hope it helps

Damien

**Comment by FabQi** [comment]
Hello

I was'nt looking in the good "Ctrlr.settings" file.
In the standalone (my) instance the "statedata" are recorded regularly and (perhaps) when you close it.

It's working / resolved

Thank you to all :-)
Fabrice
